### PR TITLE
method #element should return an element

### LINKED ISSUE #20: NoMethodError: undefined method `type' for LolSoap::WSDL::NullType
```
Exception: NoMethodError: undefined method `type' for #<LolSoap::WSDL::NullType:0x0000000545c040>
--
  0: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:27:in `block in children_hash'
  1: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:187:in `block in each'
  2: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:186:in `upto'
  3: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:186:in `each'
  4: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:25:in `children_hash'
  5: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:14:in `output'
  6: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:27:in `block in children_hash'
  7: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:187:in `block in each'
  8: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:186:in `upto'
  9: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:186:in `each'
 10: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:25:in `children_hash'
 11: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:14:in `output'
 12: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:27:in `block in children_hash'
 13: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:187:in `block in each'
 14: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:186:in `upto'
 15: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:186:in `each'
 16: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:25:in `children_hash'
 17: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:14:in `output'
 18: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:27:in `block in children_hash'
 19: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:187:in `block in each'
 20: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:186:in `upto'
 21: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:186:in `each'
 22: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:25:in `children_hash'
 23: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:14:in `output'
 24: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:27:in `block in children_hash'
 25: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:187:in `block in each'
 26: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:186:in `upto'
 27: /home/avalon/.gem/ruby/gems/nokogiri-1.6.8/lib/nokogiri/xml/node_set.rb:186:in `each'
 28: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:25:in `children_hash'
 29: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/hash_builder.rb:14:in `output'
 30: /home/avalon/.gem/ruby/gems/lolsoap-0.8.2/lib/lolsoap/response.rb:38:in `body_hash'
 31: (pry):38:in `block in <top (required)>'
```
raw response:
```
<?xml version="1.0" encoding="UTF-8"?><soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"><soapenv:Body><getWorkItemByIdResponse xmlns="http://ws.polarion.com/TrackerWebService-impl"><getWorkItemByIdReturn uri="subterra:data-service:objects:/default/OSE${WorkItem}OSE-9166" unresolvable="false" xsi:type="ns3:WorkItem" xmlns:ns1="http://ws.polarion.com/types" xmlns:ns2="http://ws.polarion.com/ProjectWebService-types" xmlns:ns3="http://ws.polarion.com/TrackerWebService-types"><ns3:approvals/><ns3:assignee/><ns3:attachments/><ns3:author uri="subterra:data-service:objects:/default/${User}user" unresolvable="false"><ns2:disabledNotifications>false</ns2:disabledNotifications><ns2:email>user@example.com</ns2:email><ns2:id>user</ns2:id><ns2:name>User Name (user)</ns2:name><ns2:voteURIs/><ns2:watcheURIs/></ns3:author><ns3:autoSuspect>false</ns3:autoSuspect><ns3:categories><ns3:Category uri="subterra:data-service:objects:/default/OSE${Category}devexp" unresolvable="false"><ns3:id>devexp</ns3:id><ns3:name>Developer Experience</ns3:name></ns3:Category></ns3:categories><ns3:comments/><ns3:created>2016-05-16T06:54:39.531Z</ns3:created><ns3:description><ns1:type>text/html</ns1:type><ns1:content>Test Tags&#xA0;</ns1:content><ns1:contentLossy>false</ns1:contentLossy></ns3:description><ns3:externallyLinkedWorkItems/><ns3:hyperlinks/><ns3:id>OSE-9166</ns3:id><ns3:linkedRevisions/><ns3:linkedRevisionsDerived/><ns3:linkedWorkItems/><ns3:linkedWorkItemsDerived/><ns3:location>default:/Products/ApplicationPlatform/OpenShift/OpenShiftEnterprise/.polarion/tracker/workitems/9000-9999/9100-9199/OSE-9166/workitem.xml</ns3:location><ns3:planningConstraints/><ns3:priority><ns3:id>50.0</ns3:id></ns3:priority><ns3:project uri="subterra:data-service:objects:/default/OSE${Project}OSE" unresolvable="false"><ns2:active>true</ns2:active><ns2:description><ns1:type>text/plain</ns1:type><ns1:content>This text comes from the project's Description in the Administration perspective. </ns1:content><ns1:contentLossy>false</ns1:contentLossy></ns2:description><ns2:id>OSE</ns2:id><ns2:lead uri="subterra:data-service:objects:/default/${User}user" unresolvable="false"><ns2:disabledNotifications>false</ns2:disabledNotifications><ns2:email>user@example.com</ns2:email><ns2:id>user</ns2:id><ns2:name>User Name (user)</ns2:name><ns2:voteURIs/><ns2:watcheURIs/></ns2:lead><ns2:location>default:/Products/ApplicationPlatform/OpenShift/OpenShiftEnterprise/.polarion/polarion-project.xml</ns2:location><ns2:name>OpenShiftEnterprise</ns2:name><ns2:projectGroupURI>subterra:data-service:objects:/default/--Products--ApplicationPlatform--OpenShift${ProjectGroup}Group</ns2:projectGroupURI><ns2:start>2015-07-20</ns2:start><ns2:trackerPrefix>OSE</ns2:trackerPrefix></ns3:project><ns3:severity><ns3:id>should_have</ns3:id></ns3:severity><ns3:status><ns3:id>draft</ns3:id></ns3:status><ns3:title>Test Tags Examples</ns3:title><ns3:type><ns3:id>testcase</ns3:id></ns3:type><ns3:updated>2016-05-16T07:23:28.183Z</ns3:updated><ns3:workRecords/><ns3:customFields><ns3:Custom><ns3:key>casecomponent</ns3:key><ns3:value xsi:type="ns3:EnumOptionId"><ns3:id xsi:type="xsd:string">auth</ns3:id></ns3:value></ns3:Custom><ns3:Custom><ns3:key>caseimportance</ns3:key><ns3:value xsi:type="ns3:EnumOptionId"><ns3:id xsi:type="xsd:string">medium</ns3:id></ns3:value></ns3:Custom><ns3:Custom><ns3:key>caselevel</ns3:key><ns3:value xsi:type="ns3:EnumOptionId"><ns3:id xsi:type="xsd:string">integration</ns3:id></ns3:value></ns3:Custom><ns3:Custom><ns3:key>caseautomation</ns3:key><ns3:value xsi:type="ns3:EnumOptionId"><ns3:id xsi:type="xsd:string">notautomated</ns3:id></ns3:value></ns3:Custom><ns3:Custom><ns3:key>caseposneg</ns3:key><ns3:value xsi:type="ns3:EnumOptionId"><ns3:id xsi:type="xsd:string">positive</ns3:id></ns3:value></ns3:Custom><ns3:Custom><ns3:key>testtype</ns3:key><ns3:value xsi:type="ns3:EnumOptionId"><ns3:id xsi:type="xsd:string">functional</ns3:id></ns3:value></ns3:Custom><ns3:Custom><ns3:key>subtype1</ns3:key><ns3:value xsi:type="ns3:EnumOptionId"><ns3:id xsi:type="xsd:string">-</ns3:id></ns3:value></ns3:Custom><ns3:Custom><ns3:key>tags</ns3:key><ns3:value xsi:type="xsd:string">ruby, mysql</ns3:value></ns3:Custom><ns3:Custom><ns3:key>products</ns3:key><ns3:value xsi:type="ns3:ArrayOfEnumOptionId"><ns3:EnumOptionId xsi:type="ns3:EnumOptionId"><ns3:id xsi:type="xsd:string">OCP, OSO</ns3:id></ns3:EnumOptionId></ns3:value></ns3:Custom></ns3:customFields></getWorkItemByIdReturn></getWorkItemByIdResponse></soapenv:Body></soapenv:Envelope>
```

WSDL is here: http://almdemo.polarion.com/polarion/ws/services/TrackerWebService?wsdl

### PR DESCRIPTION
**Pull request by akostadinov**
this fixes #20

### CODE REVIEW

**Comment by akostadinov** [comment]
bump, any chances this can be merged and a new gem release performed?

**Comment by eugenebolshakov** [comment]
I'm afraid we can't merge this as the actual problem is not with this code, but the fact is that the WSDL that you'd like to use doesn't seem to be properly supported by lolsoap (as explained in #20). We'd welcome a PR for WSDL parser to support your WSDL though.

**Comment by akostadinov** [comment]
Regardless of my original problem, this PR fixes the object type returned by the `#element` method. It currently returns type which makes no sense.

btw I don't know if the WSDL is correct or not [1]. Just have to use the API and I don't have control over the software I need to use. With this PR most everything starts to work and the method starts to make sense returning proper object type.

[1] I'm actually pretty sure WSDL is not entirely correct because found a call that I don't see description of (even in SoapUI) but performing that call does work.

**Comment by eugenebolshakov** [comment]
Again, I didn't explain this well, I'm sorry. 

A `NullType` object is returned when we try to access an element that lolsoap's WSDL parser knows nothing about. Perhaps it doesn't make sense to return `NullType.new` in `NullType#element`, it should probably just raise an error as we can't know anything about elements of an uknown type. 

Making it return a `NullElement` would be like pretending that everything is fine and making it seem that we can rely on a WSDL that we can't properly parse. It could cause other unexpected issues elsewhere. 

TLDR: the current code is confusing, but I think that merging this fix can cause even more confusion and that's why I'm reluctant to do it.

As suggested in #20 I think that not using `response.body_hash` and just parsing `response.body` using xpath is probably a better workaround.

Hope this makes sense.

**Comment by Selenita** [comment]
Hey, I have one question, I have a wsdl private, is it easy to change the link http://example.com/stockquote.wsdl for   http://MY_WSDL.wsdl  ?

**Comment by akostadinov** [comment]
I don't fully agree with the above assessment. For me hash builder should just traverse elements and get the strings no matter if type is known or unknown. Just much more convenient and low risk IMO. I think the API I have to use is too stupid to be worth the time trying to figure. And even if I figure, as I probably mentioned, there might still be non-described elements by WSDL as it seems to me looking with SoapUI.

So my proposal to project owners is to accept the patches to make unknown elements still allow to be converted to hash elements. My dirty solution is monkey patching NullElement like this:
```
class LolSoap::WSDL
  class NullType

    # https://github.com/loco2/lolsoap/pull/21
    def element(name)
      NullElement.new
    end

    def element_prefix(name)
        self.prefix
    end

    def sub_type(name)
        element(name).type
    end

    def has_attribute?(name)
        false
    end
  end
end
```

Turned out a few more things were needed than I thought originally. Leaving to the discretion of project owners and the reader to decide what to do with the information above (if anything at all).

Cheers.
